In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
import scanpy as sc
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [2]:
adata = sc.read("/data/mr423/project/data/3-OLINK_data_train_withOutlier_all.h5ad")
adata_test = sc.read("/data/mr423/project/data/3-OLINK_data_test_withOutlier_all.h5ad")

In [3]:
adata.obs

sex  DoB_Year  DoB_Month  DoB_Day         DoB Date_Attend        age  \
Id                                                                              
2144829    0      1939          1       15  1939-01-15  2007-11-16  68.835044   
3154285    0      1945          1       15  1945-01-15  2007-07-20  62.507871   
1679423    1      1945         11       15  1945-11-15  2009-05-19  63.507187   
1172610    1      1941         12       15  1941-12-15  2009-09-23  67.772758   
4011532    1      1954          1       15  1954-01-15  2009-10-20  55.761807   
...      ...       ...        ...      ...         ...         ...        ...   
1137580    1      1957          5       15  1957-05-15  2007-08-14  50.247775   
3378384    0      1950          4       15  1950-04-15  2009-06-27  59.200548   
1220136    0      1945          6       15  1945-06-15  2008-11-11  63.408624   
4988172    0      1956          1       15  1956-01-15  2010-07-13  54.491444   
5677442    1      1963          5       15  1963-05-15  2010-06-03  47.052704   

        Age_Group  
Id                 
2144829     60-70  
3154285     60-70  
1679423     60-70  
1172610     60-70  
4011532     50-60  
...           ...  
1137580     50-60  
3378384     50-60  
1220136     60-70  
4988172     50-60  
5677442     40-50  

[37304 rows x 8 columns]

In [4]:
adata_test.obs

sex  DoB_Year  DoB_Month  DoB_Day         DoB Date_Attend        age  \
Id                                                                              
4365741    0      1947         12       15  1947-12-15  2008-05-28  60.451745   
4192150    1      1951          3       15  1951-03-15  2010-01-26  58.869268   
1847588    0      1948          1       15  1948-01-15  2010-02-13  62.080767   
2209343    0      1943          3       15  1943-03-15  2009-11-24  66.696783   
4713095    1      1962          3       15  1962-03-15  2006-06-10  44.238193   
...      ...       ...        ...      ...         ...         ...        ...   
2503594    0      1947          2       15  1947-02-15  2009-08-13  62.491444   
3494250    1      1945         12       15  1945-12-15  2009-11-10  63.904175   
5746191    1      1951         12       15  1951-12-15  2009-06-26  57.530459   
4342815    0      1942          8       15  1942-08-15  2008-03-11  65.571526   
6020978    1      1947         12       15  1947-12-15  2009-07-13  61.577002   

        Age_Group  
Id                 
4365741     60-70  
4192150     50-60  
1847588     60-70  
2209343     60-70  
4713095     40-50  
...           ...  
2503594     60-70  
3494250     60-70  
5746191     50-60  
4342815     60-70  
6020978     60-70  

[4145 rows x 8 columns]

In [5]:
data = pd.read_csv("/data/mr423/project/data/3-OLINK_data_train_all_norm.csv")
data_test = pd.read_csv("/data/mr423/project/data/3-OLINK_data_test_all_norm.csv")

In [6]:
data=data.set_index('Id')
data_test=data_test.set_index('Id')

In [7]:
data

EIF4EBP1    EIF4G1     EIF5A      ENAH       ENG     ENPP2     ENPP5  \
Id                                                                              
2144829  0.156103  0.335547  0.130100  0.312544  0.369321  0.394242  0.256869   
3154285  0.199054  0.219478  0.149049  0.246452  0.450077  0.638855  0.354410   
1679423  0.278870  0.456557  0.126842  0.204109  0.371411  0.340023  0.244802   
1172610  0.373436  0.208746  0.147052  0.273067  0.336058  0.224401  0.200446   
4011532  0.357472  0.517743  0.163302  0.211867  0.412568  0.374036  0.354062   
...           ...       ...       ...       ...       ...       ...       ...   
1137580  0.327009  0.399850  0.150827  0.258535  0.394191  0.350602  0.364919   
3378384  0.251506  0.102156  0.152737  0.183943  0.455254  0.386770  0.324112   
1220136  0.200175  0.254228  0.190802  0.166639  0.397900  0.369949  0.240567   
4988172  0.199085  0.395712  0.151444  0.191822  0.389271  0.359196  0.342898   
5677442  0.262193  0.431128  0.127358  0.242793  0.391608  0.293109  0.453913   

            ENPP7    ENTPD5     EGLN1  ...    CHCHD6       CHM      CHP1  \
Id                                     ...                                 
2144829  0.120990  0.572979  0.133999  ...  0.190428  0.267989  0.075229   
3154285  0.444193  0.496786  0.158452  ...  0.247382  0.136232  0.000000   
1679423  0.000000  0.502113  0.132524  ...  0.233903  0.230272  0.106472   
1172610  0.456935  0.578152  0.204918  ...  0.187864  0.131574  0.057009   
4011532  0.149790  0.481530  0.210503  ...  0.167596  0.202356  0.138311   
...           ...       ...       ...  ...       ...       ...       ...   
1137580  0.250173  0.588624  0.174877  ...  0.190039  0.176876  0.124582   
3378384  0.206436  0.539626  0.129014  ...  0.187422  0.089019  0.112425   
1220136  0.306946  0.597404  0.117963  ...  0.156285  0.127781  0.081268   
4988172  0.101840  0.623912  0.113531  ...  0.185374  0.185091  0.112263   
5677442  0.172560  0.497766  0.175542  ...  0.179142  0.175525  0.111445   

            CHMP6    CHMP1A      CHGB      CHGA     CHRM1      KLK1     WFDC2  
Id                                                                             
2144829  0.412949  0.301072  0.344849  0.171964  0.181050  0.097833  0.312948  
3154285  0.392001  0.391468  0.231631  0.186254  0.199211  0.122182  0.291977  
1679423  0.549238  0.274063  0.236622  0.165966  0.195700  0.083087  0.236544  
1172610  0.386624  0.155575  0.286401  0.142100  0.180326  0.637118  0.282086  
4011532  0.551349  0.509913  0.290553  0.119132  0.191383  0.152645  0.000000  
...           ...       ...       ...       ...       ...       ...       ...  
1137580  0.528691  0.397152  0.227137  0.144267  0.172892  0.090240  0.213467  
3378384  0.219653  0.130143  0.244557  0.155585  0.162137  0.205746  0.245806  
1220136  0.412508  0.262803  0.276575  0.116787  0.140523  0.137852  0.231372  
4988172  0.429611  0.397983  0.283071  0.137590  0.204322  0.438541  0.218087  
5677442  0.411767  0.366274  0.200344  0.135924  0.192445  0.435615  0.199647  

[37304 rows x 2919 columns]

In [8]:
data_test

EIF4EBP1    EIF4G1     EIF5A      ENAH       ENG     ENPP2     ENPP5  \
Id                                                                              
4365741  0.162243  0.101645  0.178020  0.201666  0.418348  0.395674  0.341214   
4192150  0.381278  0.530414  0.121433  0.209079  0.418122  0.285428  0.337340   
1847588  0.182128  0.216990  0.136751  0.194922  0.379155  0.350585  0.346764   
2209343  0.152003  0.180519  0.098549  0.185611  0.391431  0.493010  0.268865   
4713095  0.249828  0.137528  0.180064  0.210663  0.452134  0.383991  0.450630   
...           ...       ...       ...       ...       ...       ...       ...   
2503594  0.263058  0.307433  0.179421  0.197109  0.471242  0.420075  0.399154   
3494250  0.236616  0.279270  0.161422  0.168190  0.396767  0.272385  0.407754   
5746191  0.199375  0.326700  0.175129  0.206027  0.399424  0.330443  0.431379   
4342815  0.179381  0.203278  0.174711  0.191245  0.412507  0.408010  0.314530   
6020978  0.232097  0.412575  0.120870  0.217232  0.390389  0.367088  0.451587   

            ENPP7    ENTPD5     EGLN1  ...    CHCHD6       CHM      CHP1  \
Id                                     ...                                 
4365741  0.256045  0.542556  0.124411  ...  0.200838  0.165568  0.092725   
4192150  0.341933  0.629422  0.207848  ...  0.194858  0.178639  0.091654   
1847588  0.228626  0.560901  0.129292  ...  0.184170  0.177902  0.112252   
2209343  0.500805  0.660234  0.116622  ...  0.170972  0.098004  0.095346   
4713095  0.228730  0.546957  0.148524  ...  0.181524  0.116203  0.113966   
...           ...       ...       ...  ...       ...       ...       ...   
2503594  0.072267  0.629179  0.163221  ...  0.178163  0.176436  0.097320   
3494250  0.315641  0.530526  0.180355  ...  0.191509  0.158802  0.098652   
5746191  0.198712  0.448008  0.134249  ...  0.178199  0.140376  0.109341   
4342815  0.243912  0.522977  0.137631  ...  0.163896  0.136275  0.076833   
6020978  0.303301  0.534506  0.170823  ...  0.237389  0.182034  0.103261   

            CHMP6    CHMP1A      CHGB      CHGA     CHRM1      KLK1     WFDC2  
Id                                                                             
4365741  0.352225  0.213849  0.301134  0.142227  0.228399  0.280694  0.234157  
4192150  0.542766  0.399148  0.320405  0.144562  0.129983  0.206868  0.225047  
1847588  0.308154  0.228718  0.333910  0.171586  0.194082  0.156126  0.229422  
2209343  0.335673  0.212658  0.246266  0.177636  0.166119  0.165108  0.264327  
4713095  0.219252  0.138960  0.347179  0.176428  0.184925  0.177714  0.323010  
...           ...       ...       ...       ...       ...       ...       ...  
2503594  0.392620  0.268021  0.281423  0.177373  0.172170  0.134185  0.435021  
3494250  0.370540  0.200643  0.239930  0.117220  0.167897  0.541750  0.225396  
5746191  0.316420  0.349413  0.212046  0.131019  0.172500  0.118112  0.178657  
4342815  0.312151  0.243827  0.309363  0.233509  0.210333  0.129994  0.220859  
6020978  0.388911  0.257717  0.346635  0.216367  0.173516  0.516268  0.290299  

[4145 rows x 2919 columns]

In [9]:
# 找出索引的交集和差异
data.index.intersection(data_test.index)

Index([], dtype='int64', name='Id')

In [11]:
y_target = adata.obs['age']
y_test = adata_test.obs['age']

In [12]:
X_train, X_val, y_train, y_val = train_test_split(data, y_target, test_size=0.2,random_state=233)

In [13]:
X_train.index.intersection(X_val.index)

Index([], dtype='int64', name='Id')

In [14]:
# 初始化并训练线性回归模型
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [18]:
y_train_pred = model.predict(X_train)

train_mse = mean_squared_error(y_train, y_train_pred)
train_rmse = np.sqrt(train_mse)
train_mae = mean_absolute_error(y_train, y_train_pred)
train_mape = np.mean(np.abs((y_train - y_train_pred) / y_train)) * 100
train_r2 = r2_score(y_train, y_train_pred)

print("train:")
print(f"MSE: {train_mse:.4f}")
print(f"RMSE: {train_rmse:.4f}")
print(f"MAE: {train_mae:.4f}")
print(f"MAPE: {train_mape:.4f}%")
print(f"R²: {train_r2:.4f}")

train:
MSE: 7.4023
RMSE: 2.7207
MAE: 2.1431
MAPE: 3.8118%
R²: 0.8898


In [19]:
y_val_pred = model.predict(X_val)

val_mse = mean_squared_error(y_val, y_val_pred)
val_rmse = np.sqrt(val_mse)
val_mae = mean_absolute_error(y_val, y_val_pred)
val_mape = np.mean(np.abs((y_val - y_val_pred) / y_val)) * 100
val_r2 = r2_score(y_val, y_val_pred)

print("val:")
print(f"MSE: {val_mse:.4f}")
print(f"RMSE: {val_rmse:.4f}")
print(f"MAE: {val_mae:.4f}")
print(f"MAPE: {val_mape:.4f}%")
print(f"R²: {val_r2:.4f}")

val:
MSE: 9.0184
RMSE: 3.0031
MAE: 2.3814
MAPE: 4.2465%
R²: 0.8650


In [20]:
y_test_pred = model.predict(data_test)

test_mse = mean_squared_error(y_test, y_test_pred)
test_rmse = np.sqrt(test_mse)
test_mae = mean_absolute_error(y_test, y_test_pred)
test_mape = np.mean(np.abs((y_test - y_test_pred) / y_test)) * 100
test_r2 = r2_score(y_test, y_test_pred)

print("test:")
print(f"MSE: {test_mse:.4f}")
print(f"RMSE: {test_rmse:.4f}")
print(f"MAE: {test_mae:.4f}")
print(f"MAPE: {test_mape:.4f}%")
print(f"R²: {test_r2:.4f}")

test:
MSE: 9.5362
RMSE: 3.0881
MAE: 2.4305
MAPE: 4.3247%
R²: 0.8551


In [12]:
seeds = [42, 0, 233, 1234, 9999]
# 选择seed = 233
for seed in seeds:
    print(f'seed = {seed}')
    X_train, X_val, y_train, y_val = train_test_split(data, y_target, test_size=0.2, random_state=seed)
    
    model = LinearRegression()
    model.fit(X_train, y_train)
    y_train_pred = model.predict(X_train)

    train_mse = mean_squared_error(y_train, y_train_pred)
    train_rmse = np.sqrt(train_mse)
    train_mae = mean_absolute_error(y_train, y_train_pred)
    train_mape = np.mean(np.abs((y_train - y_train_pred) / y_train)) * 100
    train_r2 = r2_score(y_train, y_train_pred)

    print("train:")
    print(f"MSE: {train_mse:.4f}")
    print(f"RMSE: {train_rmse:.4f}")
    print(f"MAE: {train_mae:.4f}")
    print(f"MAPE: {train_mape:.4f}%")
    print(f"R²: {train_r2:.4f}")



    y_val_pred = model.predict(X_val)

    val_mse = mean_squared_error(y_val, y_val_pred)
    val_rmse = np.sqrt(val_mse)
    val_mae = mean_absolute_error(y_val, y_val_pred)
    val_mape = np.mean(np.abs((y_val - y_val_pred) / y_val)) * 100
    val_r2 = r2_score(y_val, y_val_pred)

    print("val:")
    print(f"MSE: {val_mse:.4f}")
    print(f"RMSE: {val_rmse:.4f}")
    print(f"MAE: {val_mae:.4f}")
    print(f"MAPE: {val_mape:.4f}%")
    print(f"R²: {val_r2:.4f}")



    y_test_pred = model.predict(data_test)

    test_mse = mean_squared_error(y_test, y_test_pred)
    test_rmse = np.sqrt(test_mse)
    test_mae = mean_absolute_error(y_test, y_test_pred)
    test_mape = np.mean(np.abs((y_test - y_test_pred) / y_test)) * 100
    test_r2 = r2_score(y_test, y_test_pred)

    print("test:")
    print(f"MSE: {test_mse:.4f}")
    print(f"RMSE: {test_rmse:.4f}")
    print(f"MAE: {test_mae:.4f}")
    print(f"MAPE: {test_mape:.4f}%")
    print(f"R²: {test_r2:.4f}")


seed = 42
train:
MSE: 7.3984
MAE: 2.1496
R²: 0.8896
val:
MSE: 9.0332
MAE: 2.3417
R²: 0.8659
test:
MSE: 9.4797
MAE: 2.4124
R²: 0.8560
seed = 0
train:
MSE: 7.3941
MAE: 2.1451
R²: 0.8902
val:
MSE: 9.0692
MAE: 2.3824
R²: 0.8628
test:
MSE: 9.5118
MAE: 2.4318
R²: 0.8555
seed = 233
train:
MSE: 7.4023
MAE: 2.1431
R²: 0.8898
val:
MSE: 9.0184
MAE: 2.3814
R²: 0.8650
test:
MSE: 9.5362
MAE: 2.4305
R²: 0.8551
seed = 1234
train:
MSE: 7.4065
MAE: 2.1480
R²: 0.8894
val:
MSE: 9.0467
MAE: 2.3821
R²: 0.8664
test:
MSE: 9.5532
MAE: 2.4299
R²: 0.8549
seed = 9999
train:
MSE: 7.3375
MAE: 2.1403
R²: 0.8904
val:
MSE: 9.2958
MAE: 2.3966
R²: 0.8626
test:
MSE: 9.4120
MAE: 2.4146
R²: 0.8570


seed = 233

train:
MSE: 7.4023
MAE: 2.1431
R²: 0.8898

val:
MSE: 9.0184
MAE: 2.3814
R²: 0.8650

test:
MSE: 9.5362
MAE: 2.4305
R²: 0.8551